modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Transactions/TransactionsClient.cs

```diff
@@ -0,0 +1,257 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Apache.Ignite.Core.Impl.Client.Transactions
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Threading;
+    using Apache.Ignite.Core.Client;
+    using Apache.Ignite.Core.Client.Transactions;
+    using Apache.Ignite.Core.Impl.Binary;
+    using Apache.Ignite.Core.Impl.Common;
+    using Apache.Ignite.Core.Transactions;
+
+    /// <summary>
+    /// Ignite Thin Client transactions facade.
+    /// </summary>
+    internal class TransactionsClient : ITransactionsClient, IDisposable
+    {
+        /** Default transaction configuration. */
+        private readonly TransactionClientConfiguration _cfg;
+
+        /** Transaction for this thread and client. */
+        private readonly ThreadLocal<TransactionClient> _currentTx = new ThreadLocal<TransactionClient>();
+
+        /** Ignite. */
+        private readonly IgniteClient _ignite;
+
+        /** Transaction manager. */
+        private readonly ClientCacheTransactionManager _txManager;
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="ignite">Ignite.</param>
+        /// <param name="cfg"></param>
+        public TransactionsClient(IgniteClient ignite, TransactionClientConfiguration cfg)
+        {
+            _ignite = ignite;
+            _cfg = cfg ?? new TransactionClientConfiguration();
+            _txManager = new ClientCacheTransactionManager(this);
+        }
+
+        /** <inheritDoc /> */
+        [SuppressMessage("Microsoft.Usage", "CA1816:CallGCSuppressFinalizeCorrectly",
+            Justification = "There is no finalizer.")]
+        public void Dispose()
+        {
+            _currentTx.Dispose();
+            _txManager.Dispose();
+        }
+
+        /// <summary>
+        /// Starts ambient transaction if needed.
+        /// </summary>
+        internal void StartTxIfNeeded()
+        {
+            _txManager.StartTxIfNeeded();
+        }
+
+        /** <inheritDoc /> */
+        ITransactionClient ITransactionsClient.Tx
+        {
+            get { return Tx; }
+        }
+
+        /// <summary>
+        /// Gets transaction started by this thread or null if this thread does not have a transaction.
+        /// </summary>
+        internal TransactionClient Tx
+        {
+            get
+            {
+                var tx = _currentTx.Value;
+
+                if (tx == null)
+                    return null;
+
+                if (tx.Closed)
```

Comment:
We should handle an edge case when multiple clients are used from the same thread:
check that `tx` actually belongs to the current `IgniteClient` and has the same `ClientSocket`.

If there is an active transaction for another client or socket, I think we should throw  `NotSupportedException`.

Reply:
Also, when socket is disconnected, we should clear all related thread locals to avoid leading memory.